jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java

```diff
@@ -0,0 +1,457 @@
+            if (connectionFactory==null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no connection factory {}",baseRequest);
+                return false;
+            }
+            HttpChannel channel = baseRequest.getHttpChannel();
+            WebSocketCoreConnection connection =
+                    connectionFactory.newConnection(channel.getConnector(),channel.getEndPoint(),extensionCSVs.getValues());
+            if (connection==null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no connection {}",baseRequest);
+            }
+
+            // Create a session from a session factory (from context or connector)
+            WebSocketSessionFactory sessionFactory = null;
+            if (context!=null)
+                sessionFactory = (WebSocketSessionFactory)context.getAttribute(WebSocketSessionFactory.class.getName());
+            if (sessionFactory==null)
+                sessionFactory = baseRequest.getHttpChannel().getConnector().getBean(WebSocketSessionFactory.class);
+            if (sessionFactory==null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no session factory {}",baseRequest);
+                return false;
+            }
+            WebSocketCoreSession session = sessionFactory.newSession(connection, subprotocol);
+            if (session==null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no session {}",baseRequest);
+            }
+
+            connection.setSession(session);
+
+            // send upgrade response
+            Response baseResponse = baseRequest.getResponse();
+            baseResponse.setStatus(HttpServletResponse.SC_SWITCHING_PROTOCOLS);
+            baseResponse.getHttpFields().add(UpgradeWebSocket);
+            baseResponse.getHttpFields().add(ConnectionUpgrade);
+            baseResponse.getHttpFields().add(HttpHeader.SEC_WEBSOCKET_ACCEPT, AcceptHash.hashKey(key));
+            baseResponse.getHttpFields().add(HttpHeader.SEC_WEBSOCKET_EXTENSIONS,
+                                             ExtensionConfig.toHeaderValue(connection.getExtensionStack().getNegotiatedExtensions()));
+            if (subprotocol!=null)
+                baseResponse.getHttpFields().add(HttpHeader.SEC_WEBSOCKET_SUBPROTOCOL,subprotocol);
+            baseResponse.flushBuffer();
+            baseRequest.setHandled(true);
+
+            // upgrade
+            if (LOG.isDebugEnabled())
+                LOG.debug("upgrade connection={} session={}",connection,session);
+
+            baseRequest.setAttribute(HttpConnection.UPGRADE_CONNECTION_ATTRIBUTE, connection);
+            return true;
+        }
+    };
+
+
+
+    static class TestWebSocketConnectionFactory extends ContainerLifeCycle implements WebSocketConnectionFactory
```

Comment:
WebSocketConnectionFactory ??

That is not required.
Each implementation will have its own raw creation of a Connection.

Reply:
Why?   I can see nothing in the connection that is implementation specific.   The connection can be concrete and in core.   Only the session needs to be application layer specific